### PR TITLE
 Enable saving histograms, rework histogram grouping

### LINKED ISSUE #11733: Populate column options in histogram from data
In the DE, if I run a query while visualizing a histogram, the _X_ and _Fill_ visualization options should populate with the column names returned by the query.

### LINKED ISSUE #11735: Add histogram as generic visualization type on backend

### PR DESCRIPTION
**Pull request by chnn**
Closes #11735 
Closes #11733 

This commit does three things:

1. Enables saving histogram views to a dashboard
2. Uses real data for the histogram column options and keeps them in sync when a view's data changes
3. Reworks the histogram grouping option. Data in a histogram can be grouped by more than a single column now, and the default selection of columns is roughly the same as the group key returned in the Flux response

   grouping

### CODE REVIEW

File: ui/src/minard/utils/bin.ts [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+  const domain = extent(col)
+  const d0 = domain[0]
+
+  let d1 = domain[1]
+
+  if (d0 === d1) {
+    d1 = d0 + 1
+  }
+
+  const bins = range(d0, d1, (d1 - d0) / binCount).map(min => ({
```
Comment (ebb-tide) [inline ui/src/minard/utils/bin.ts]: This line is very concise, but also hard to parse visually. If you're thinking of this as a separate library and util functions I think it would be better to err on the side of verbosity instead. This would make it easy for the next person to look at the code to be able to reason about it. This is my general comment on this file and others in minard.

File: ui/src/minard/components/Histogram.tsx [hunk 2]
```diff
@@ -76,33 +62,22 @@ export const Histogram: SFC<Props> = props => {
...
+  const {aesthetics, table: statTable} = layer
 
   let hoveredRowIndices = null
 
   if (hoverX && hoverY) {
     hoveredRowIndices = findHoveredRowIndices(
-      xMinCol,
-      xMaxCol,
-      yMaxCol,
+      statTable.columns[aesthetics.xMin],
```
[the remark this answers is not in the document]
Comment (chnn) [reply ui/src/minard/components/Histogram.tsx]: maybe it could be called `columnMappings` or something? a bit verbose but more to the point

**Comment by chnn** [comment]
@ebb-tide I've been struggling a bit to balance ease of experimentation and code quality while working on this. I view the current code as more or less a throwaway implementation—normally not the sort of stuff I would ever commit, but I want to get validation on the general approach (what's described in the RFC) before spending the time to make it maintainable for the long haul.

I can see this code going in one of two ways:

1. There's not enough enthusiastic support for the approach sketched out here and described in the RFC. In that case I'll drop everything in `src/minard` and implement the simplest possible implementation of the histogram view type. I think it can be drastically simplified if it doesn't have to account for fitting into a general framework.

2. There is support for approach described in the RFC, in which case I'll extract much of this to a separate package and refactor along the way (likely as a pair with someone).

At the same time, this is _is_ a functional implementation of a new and shiny feature. Russ is eager to get some feedback and I certainly am as well. So maybe a third way this goes is that we end up wanting a drastically different UX and the code gets scrapped in service of that.

Regardless, there's some technical debit here I intend to clean up. Sorry I didn't communicate that too well before opening up the PR, but I appreciate you slogging through nonetheless!

File: ui/src/minard/utils/getBarFill.ts [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+export const getBarFill = (
+  {scales, aesthetics, table}: Layer,
+  i: number
+): string => {
+  const fillScale = scales.fill
+  const values = aesthetics.fill.map(colKey => table.columns[colKey][i])
+  const groupKey = getGroupKey(values)
+  const fill = fillScale(groupKey)
+
```
[the remark this answers is not in the document]
Comment (chnn) [reply ui/src/minard/utils/getBarFill.ts]: Can do!

File: ui/src/clockface/components/dropdowns/MultiSelectDropdown.tsx [hunk 4]
```diff
@@ -252,9 +252,12 @@ class MultiSelectDropdown extends Component<Props, State> {
...
 
   private validateChildCount = (): void => {
-    const {children} = this.props
+    const {children, status} = this.props
 
-    if (React.Children.count(children) === 0) {
+    const shouldHaveChildren =
+      status === ComponentStatus.Default || status === ComponentStatus.Valid
+
+    if (shouldHaveChildren && React.Children.count(children) === 0) {
```
[the remark this answers is not in the document]
Comment (chnn) [reply ui/src/clockface/components/dropdowns/MultiSelectDropdown.tsx]: @alexpaxton @mavarius The interface for the `MultiSelectDropdown` drifted a bit from the `Dropdown`. This was one change made only in the `Dropdown`.

Another, that I didn't update here, was that the `MultiSelectDropdown` expects the `value` prop of a `Dropdown.Item` to be an object with an `id` field. It will call the `onChange` handler with the `id`. This is different than the `Dropdown` component, which will call the `onChange` handler with whatever is passed to the `value` prop of each `Dropdown.Item`.

**Comment by ebb-tide** [comment]
I definitely understand that, I wasn't clear how settled you felt this feature was, and I enthusiastically support tech debt in favor or working feature. It would be super cool if you could pair as you finalize this library.
